### PR TITLE
dev Local Bundle Adjustment

### PR DESCRIPTION
**Pull request by Debize**
Summary:
- A. Local Bundle Adjustment
- B. Maximize the number of loaded matches

----------------------------------------------------------------------------
#			 A. Local Bundle Adjustment
----------------------------------------------------------------------------

It's time to lose yourself in the wonderful world of the "Local Bundle Adjustment".

The Local BA (LBA) approach allows to decrease drastically the reconstruction time reducing the number of parameters given to the Ceres Solver.

## -- How does it work ? 

Usually, during a classical Incremental SfM pipeline, after adding a group of N new views to the reconstructed scene we perform a Bundle Adjustment (BA) to adjust ALL the parameters of the scene (poses, landmarks & intrinsics).
With this method, the BA part becomes increasingly time-consuming. 

Our method attempts to reduce the number of parameters to adjust, using the following paradigm: 

_"Let adjust the parameters close to the N newly resected cameras only."_ 

(Considering the reconstruction of a long street, when new views are added at the end of it the BA won't modify significantly the parameters located at the beginning of it.)

Before performing a BA, we build a graph in which the views form the Nodes, and an edge is created if 2 views share at least M matches.

Then, the graph is split into 3 regions :
- The “_active region_” (**RA**) includes all the parameters being REFINED by the solver.
- The “_passive region_” (**RP**) includes all the parameters set to CONSTANT in the solver.
- The “_skipped region_” (**RS**) refers to the other parameters, NOT INCLUDED in the solver.

To define these regions, we perform a breadth-first search in the graph using the newly resected cameras as source. This technique gives us the distance from each camera of the graph to the newly resected cameras. We use a graph-distance threshold D (e.g. D=1) to define those regions:
- a view belongs to RA <=> its distance is <= D 
- a view belongs to RP <=> its distance is == D + 1
- a view belongs to RS <=> its distance is > D + 1
- a landmark belongs to RA <=> if at least 1 of these observations belongs to RA
- a landmark belongs to RS <=> it does not belong to RA
- an intrinsic belongs to RA <=> the variation of its focal length is < a threshold.
- an intrinsic belongs to RP <=> it does not belong to RA

## -- Performances

The algorithm has been tested on 7 datasets with different configurations (sources: Mikros, Popart, Tanks&Temples) including from 371 to 1106 pictures and compared to the classic BA.

The speed gain varies from a factor 3.4 to 15.9.

The number of reconstructed points varies from 93% to 107%
The number of reconstructed cameras varies from 95% to 104%

## -- How to use it ? 

The Local BA can only be used with in the `main_incrementalSfM` executable.
You only need to add the argument **--useLocalBA 1** (or **-l 1**) to enable it.

## -- Implementation

**sfm/LocalBundleAdjustmentData.\*pp**

`LocalBundleAdjustmentData` includes the distance-graph, all the maps containing the state (refined, constant or ignored) of all the parameters before each refinement and the function `updateParametersState(...)` establishing the Local BA context according to the current reconstruction scene.

**sfm/LocalBundleAdjustmentCeres.\*pp**

`LocalBundleAdjustmentCeres` inherits from the `BundleAdjustmentCeres` class (for the `Adjust()` function). It contains all that you need to transfer the reconstruction parameters to the Ceres Sover according to the specified rules included in `LocalBundleAdjustmentData`.

**sfm/bundleAdjustment\_test.cpp**

It contains a test about Local BA (named `LOCAL_BUNDLE_ADJUSTMENT_EffectiveMinimization_Pinhole_CamerasRing`) providing a exemple of the Local BA use.

----------------------------------------------------------------------------
#		B. Maximize the number of loaded matches
----------------------------------------------------------------------------

With the aim of perform a fast reconstruction (e.g. during a demo), we add an option to the `main_incrementalSfM` to curb the number of loaded matches (for each type of descriptor).

## -- Performances

*in progress*

## -- How to use it ? 

Add the argument **--maxNumberOfMatches M**, with _M_ the number of matches you want to load.

### CODE REVIEW

File: src/aliceVision/sfm/LocalBundleAdjustmentData.hpp [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+  
+  /// Set the output path where Local BA outputs will be saved
+  void setOutDirectory(const std::string& dir)    {_outFolder = dir;}
+    
+  /// @brief Set every parameters of the BA problem to Refine: the Local BA becomes a classic BA.
+  /// @param[in] sfm_data contains all the data about the reconstruction.
+  void setAllParametersToRefine(const SfMData& sfm_data);
+  
+  /// @brief Add the newly resected views into a graph and use this graph to set an \c EState (refined, constant, ignored) for 
+  /// each parameter in the bundle adjutment (poses, landmarks, intrinsics).
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentData.hpp]: adjustment

File: src/aliceVision/sfm/LocalBundleAdjustmentCeres.cpp [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+  
+  int numCamDistUpperOne = 0;
+  for (const auto & camdistIt : _numCamerasPerDistance)
+  {
+    if (camdistIt.first > 1)
+      numCamDistUpperOne += camdistIt.second;
+  }
+  
+  ALICEVISION_LOG_DEBUG("\n----- Local BA Ceres statistics ------\n" 
+                        << "|- adjutment duration: " << _time << " s \n"
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentCeres.cpp]: adjustment

File: src/aliceVision/sfm/LocalBundleAdjustmentCeres.cpp [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+    header.push_back("ResidualBlocks");
+    header.push_back("SuccessIter"); header.push_back("BadIter");
+    header.push_back("InitRMSE"); header.push_back("FinalRMSE");
+    header.push_back("dAR=-1");
+    header.push_back("dAR=0"); header.push_back("dAR=1"); header.push_back("dAR=2");
+    header.push_back("dAR=3"); header.push_back("dAR=4"); header.push_back("dAR=5");   
+    header.push_back("dAR=6"); header.push_back("dAR=7"); header.push_back("dAR=8"); 
+    header.push_back("dAR=9"); header.push_back("dAR=10+");
+    header.
+        push_back("New Views");
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentCeres.cpp]: formatting

File: src/aliceVision/sfm/LocalBundleAdjustmentData.cpp [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+    }
+  }
+  else // the graph is not empty
+    addedViewsId = newReconstructedViews;
+  
+  std::map<Pair, std::size_t> map_imagesPair_nbSharedLandmarks 
+      = countSharedLandmarksPerImagesPair(sfm_data, map_tracksPerView, addedViewsId);
+  
+  // -- Add nodes to the graph
+  std::vector<lemon::ListGraph::Node> newNodes;
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentData.cpp]: unused

File: src/aliceVision/sfm/LocalBundleAdjustmentData.cpp [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+  }
+  
+  // -- Add edges to the graph
+  std::size_t numEdges = 0;
+  
+  for(const auto& x: map_imagesPair_nbSharedLandmarks)
+  {
+    if(x.second > kMinNbOfMatches) // ensure a minimum number of landmarks in common to consider the link
+    {
+      _graph.addEdge(_mapNodePerViewId.at(x.first.first), _mapNodePerViewId.at(x.first.second));
```
Comment (fabiencastan) [inline src/aliceVision/sfm/LocalBundleAdjustmentData.cpp]: Problem with keys in _mapNodePerViewId here

File: src/aliceVision/sfm/LocalBundleAdjustmentCeres.cpp [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+      numCamDistUpperOne += camdistIt.second;
+  }
+  
+  ALICEVISION_LOG_DEBUG("\n----- Local BA Ceres statistics ------\n" 
+                        << "|- adjutment duration: " << _time << " s \n"
+                        << "|- graph-distances distribution: \n"
+                        << "  * not connected: " << _numCamerasPerDistance[-1] << " cameras \n"
+                        << "  * D = 0: " << _numCamerasPerDistance[0] << " cameras \n"
+                        << "  * D = 1: " << numCamDistEqOne << " cameras \n"
+                        << "  * D > 1: " << numCamDistUpperOne << " cameras \n"
```
[the remark this answers is not in the document]
Comment (Debize) [reply src/aliceVision/sfm/LocalBundleAdjustmentCeres.cpp]: As a parameter of **_ReconstructionEngine_sequentialSfM::localBundleAdjustment()_**

File: src/software/pipeline/main_incrementalSfM.cpp [hunk 7]
```diff
@@ -137,7 +138,9 @@ int main(int argc, char **argv)
       "If the process is done on renderfarm, it doesn't make sense to wait for user inputs")
     ("useLocalBA,l", po::value<bool>(&useLocalBundleAdjustment)->default_value(useLocalBundleAdjustment),
       "Enable/Disable the Local bundle adjustment strategy.\n"
-      "It reduces the reconstruction time, especially for big datasets (500+ images).\n");
+      "It reduces the reconstruction time, especially for big datasets (500+ images).\n")
+    ("localBAGraphDistance", po::value<std::size_t>(&localBundelAdjustementGraphDistanceLimit)->default_value(localBundelAdjustementGraphDistanceLimit),
+      "Graph-distance limit setting the Active region in the Local Bundle Adjustment strategy (by default: 1).\n");
```
Comment (fabiencastan) [inline src/software/pipeline/main_incrementalSfM.cpp]: Proposition:
"Graph-distance limit to define the Active Region in the Local Bundle Adjustment strategy."

**Pull request merged**